admittance_controller/include/admittance_controller/admittance_rule_impl.hpp

```diff
@@ -0,0 +1,444 @@
+  }
+
+  bool success = get_all_transforms(current_joint_state, reference_joint_state, admittance_state_);
+
+  // calculate needed rotations
+  Eigen::Matrix<double, 3, 3> rot_base_sensor = trans_.base_sensor_.rotation();
+  Eigen::Matrix<double, 3, 3> rot_world_base = trans_.world_base_.rotation();
+  Eigen::Matrix<double, 3, 3> rot_base_cog = trans_.base_cog_.rotation();
+  Eigen::Matrix<double, 3, 3> rot_base_control = trans_.base_control_.rotation();
+
+  // apply filter and update wrench_world_ vector
+  Eigen::Matrix<double, 3, 3> rot_world_sensor = rot_world_base * rot_base_sensor;
+  Eigen::Matrix<double, 3, 3> rot_world_cog = rot_world_base * rot_base_cog;
+  process_wrench_measurements(measured_wrench, rot_world_sensor, rot_world_cog);
+
+  // transform wrench_world_ into base frame
+  admittance_state_.wrench_base.block<3, 1>(0, 0) =
+    rot_world_base.transpose() * wrench_world_.block<3, 1>(0, 0);
+  admittance_state_.wrench_base.block<3, 1>(3, 0) =
+    rot_world_base.transpose() * wrench_world_.block<3, 1>(3, 0);
+
+  // Compute admittance control law
+  vec_to_eigen(current_joint_state.positions, admittance_state_.current_joint_pos);
+  admittance_state_.rot_base_control = rot_base_control;
+  admittance_state_.ref_trans_base_ft = trans_ref_.base_ft_;
+  admittance_state_.ft_sensor_frame = parameters_.ft_sensor.frame.id;
+  success &= calculate_admittance_rule(admittance_state_, dt);
+
+  // if a failure occurred during any kinematics interface calls, return an error and don't
+  // modify the desired reference
+  if (!success)
+  {
+    desired_joint_state = reference_joint_state;
+    return controller_interface::return_type::ERROR;
+  }
+
+  // update joint desired joint state
+  for (size_t i = 0; i < reference_joint_state.positions.size(); ++i)
+  {
+    desired_joint_state.positions[i] =
+      reference_joint_state.positions[i] + admittance_state_.joint_pos[i];
+  }
+  for (size_t i = 0; i < reference_joint_state.velocities.size(); ++i)
+  {
+    desired_joint_state.velocities[i] = admittance_state_.joint_vel[i];
+  }
+  for (size_t i = 0; i < reference_joint_state.accelerations.size(); ++i)
+  {
+    desired_joint_state.accelerations[i] = admittance_state_.joint_acc[i];
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+bool AdmittanceRule::calculate_admittance_rule(AdmittanceState & admittance_state, double dt)
+{
+  // create stiffness matrix
+  auto rot_base_control = admittance_state.rot_base_control;
+  Eigen::Matrix<double, 6, 6> K = Eigen::Matrix<double, 6, 6>::Zero();
+  Eigen::Matrix<double, 3, 3> K_pos = Eigen::Matrix<double, 3, 3>::Zero();
+  Eigen::Matrix<double, 3, 3> K_rot = Eigen::Matrix<double, 3, 3>::Zero();
+  K_pos.diagonal() = admittance_state.stiffness.block<3, 1>(0, 0);
+  K_rot.diagonal() = admittance_state.stiffness.block<3, 1>(3, 0);
+  K_pos = rot_base_control * K_pos * rot_base_control.transpose();
+  K_rot = rot_base_control * K_rot * rot_base_control.transpose();
```

Comment:
Are you here rotating stiffness into control frame? Why?

If stiffness is defined in 6D, it should be applied in the control frame all the time. Or are you always doing control in base frame. If so, why?

Why not have always everything in control frame here and simply calculate everything?

Reply:
The Jacobin and transforms are expressed in the base_link frame by default. The kinematics interface does not support converting Cartesian velocities to joint velocities in an arbitrary frame.

Reply:
Ok, so this means that you have everything in "base_link" and then you are converting from control frame to base link?
Having few comment explaining the process would be great. So one doesn't have to actually read the code in detail.